Blog post: Import Path FAQ (#147)
ethereum/solidity-blog: _posts/2021-08-16-import-path-faq.md

title: 'Solidity Import Path FAQ'
date: '2021-08-16'
author: Kamil Śliwak
category: Explainers
---

Paths are one of the things that "just work" as long as you're not too particular about the details.
If you can just hand off the path to a library, it's no longer your problem.
And the details can get complicated if your program has to work on a variety of platforms with completely different filesystem conventions or even not having a filesystem at all (e.g. a browser).

This is the very reason why the Solidity compiler takes a hands-off approach to the problem.
The [Remix IDE](https://remix.ethereum.org/) is widely known for supporting Github URLs but what many developers do not know is that you can use literally any string to identify your file.
Internally the compiler makes very few assumptions about that string which, ironically, sometimes leads to misunderstandings when it clashes with programmer's expectations.
In the end the source code in most cases comes from a filesystem and users do not refer to it using abstract identifiers.
They expect the compiler to map their local filesystem content to its own internal structure in a consistent way regardless of the platform.

> Even when using [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format, allows tools to access that structure directly, the burden just shifts to the tool.

Reviewer: grammar is off

---
layout: post
published: true
title: 'Solidity Import Path FAQ'
date: '2021-08-16'
author: Kamil Śliwak
category: Explainers
---

Paths are one of the things that "just work" as long as you're not too particular about the details.
If you can just hand off the path to a library, it's no longer your problem.
And the details can get complicated if your program has to work on a variety of platforms with completely different filesystem conventions or even not having a filesystem at all (e.g. a browser).

This is the very reason why the Solidity compiler takes a hands-off approach to the problem.
The [Remix IDE](https://remix.ethereum.org/) is widely known for supporting Github URLs but what many developers do not know is that you can use literally any string to identify your file.
Internally the compiler makes very few assumptions about that string which, ironically, sometimes leads to misunderstandings when it clashes with programmer's expectations.

> In the end the source code in most cases comes from a filesystem and users do not refer to it using abstract identifiers.

Reviewer 2: ```suggestion
In the end the source code comes in most cases from a filesystem and users don't refer to it using abstract identifiers.
```

They expect the compiler to map their local filesystem content to its own internal structure in a consistent way regardless of the platform.

> Even when using [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format, allows tools to access that structure directly, the burden just shifts to the tool.

Reviewer 2: ```suggestion
And while our [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format allows tools to access that structure directly, the burden just shifts to the tool.
```

> The details of the whole mechanism have recently been documented in [Import Path Resolution](https://docs.soliditylang.org/en/latest/path-resolution.html) but many of the consequences of the documented behavior may not be apparent.

Reviewer 2: ```suggestion
The details of the whole mechanism have recently been added to our [Import Path Resolution documentation](https://docs.soliditylang.org/en/latest/path-resolution.html) but many of the consequences of the documented behavior may not be apparent.
```

This blog post is meant to be a more complete exploration of problems you might encounter when dealing with paths in Solidity.
It will show you specific corner cases, explain why they happen and tell what you can do to avoid them.

> ### Q: What's wrong with relative paths in Solidity? Aren't `import "file.sol"` and `import "./file.sol"` the same thing?

Reviewer 2: ```suggestion
### Q: What's wrong with relative paths in Solidity? Isn't `import "file.sol"` the same as `import "./file.sol"`?
```

Yours is grammatically correct as well. Dunno I felt mine reads nicer, but just a suggestion.

No, in Solidity these two statements are processed very differently.

> `file.sol` is actually considered absolute in compiler's virtual filesystem.

Reviewer 2: ```suggestion
`file.sol` is actually considered absolute in the compiler's virtual filesystem.
```
